# 01 Importing Libraries

In [3]:
import pandas as pd
import numpy as np
import os

# 02 Importing Data

In [4]:
path = r'/Users/woodoooo/Desktop/Instacart Basket Analysis/'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled_task_4.4.csv'), index_col = False)

# 03 Descriptive Statistics & Exploratory analysis for products.csv

In [33]:
# Checking for mixed types 
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].map(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

product_name


In [37]:
df_prods.dtypes

product_id         int64
product_name      object
aisle_id           int64
department_id      int64
prices           float64
dtype: object

In [5]:
# Checking missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [18]:
# Creating a datafarme to subset the data in the df_prods where there  missing values are found in the product_name column

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [7]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [8]:
df_prods.shape

(49693, 5)

In [9]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [10]:
df_prods_clean.shape

(49677, 5)

In [11]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [12]:
df_prods_clean.duplicated().sum()

np.int64(5)

In [13]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [14]:
df_prods_clean.shape

(49677, 5)

In [15]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [16]:
df_prods_clean_no_dups.shape

(49672, 5)

In [19]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [21]:
#Fidning the median of the df_prods dataframe
df_prods_clean_no_dups['prices'].median() 

7.1

Prices Column:

Median price: 7.1 (typical product price).

Max price: 99,999 — this is clearly unrealistic.

Mean price: ~9.99, but this is skewed by the outlier.

Most prices (75%) are below 11.1

In [27]:
#Finding outliers in the df_prods dataframe.

# Step 1: Calculate threshold for outliers
price_mean = df_prods_clean_no_dups['prices'].mean()
price_std = df_prods_clean_no_dups['prices'].std()
threshold = price_mean + (2 * price_std)

# Step 2: Filter rows where prices are above the threshold (outliers)
outliers = df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] > threshold]

# Step 3: Display the outliers
outliers

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [28]:
# Replace outliers with median price
df_prods_clean_no_dups.loc[
    df_prods_clean_no_dups['prices'] > threshold,
    'prices'
] = 7.1

In [29]:
#Checking the df_prods_clean_no_dups dataframe after imputation
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      7.680414
std    14340.705287     38.315784       5.850779      4.199298
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

In [39]:
#Checking the number of rows & columns in thedataframe
df_prods_clean_no_dups.shape

(49672, 5)

# 04 Exporting Data

In [38]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))